mpt/impl/smtp/cassandra-rabbitmq-awss3/src/test/java/org/apache/james/mpt/smtp/CassandraRabbitMQAwsS3SmtpTestRule.java

```diff
@@ -0,0 +1,184 @@
+import org.apache.james.modules.blobstore.BlobStoreChoosingModule;
+import org.apache.james.modules.protocols.ProtocolHandlerModule;
+import org.apache.james.modules.protocols.SMTPServerModule;
+import org.apache.james.modules.protocols.SmtpGuiceProbe;
+import org.apache.james.modules.rabbitmq.RabbitMQModule;
+import org.apache.james.modules.server.CamelMailetContainerModule;
+import org.apache.james.mpt.api.Continuation;
+import org.apache.james.mpt.api.Session;
+import org.apache.james.mpt.monitor.SystemLoggingMonitor;
+import org.apache.james.mpt.session.ExternalSessionFactory;
+import org.apache.james.queue.api.MailQueueItemDecoratorFactory;
+import org.apache.james.queue.api.RawMailQueueItemDecoratorFactory;
+import org.apache.james.server.core.configuration.Configuration;
+import org.apache.james.util.Host;
+import org.apache.james.util.Port;
+import org.apache.james.utils.DataProbeImpl;
+import org.junit.rules.TemporaryFolder;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Splitter;
+import com.google.inject.Module;
+import com.google.inject.util.Modules;
+
+public class CassandraRabbitMQAwsS3SmtpTestRule implements TestRule, SmtpHostSystem {
+
+    enum SmtpServerConnectedType {
+        SMTP_GLOBAL_SERVER(probe -> Port.of(probe.getSmtpPort())),
+        SMTP_START_TLS_SERVER(probe -> Port.of(probe.getSmtpsPort()));
+
+        private final Function<SmtpGuiceProbe, Port> portExtractor;
+
+        SmtpServerConnectedType(Function<SmtpGuiceProbe, Port> portExtractor) {
+            this.portExtractor = portExtractor;
+        }
+
+        public Function<SmtpGuiceProbe, Port> getPortExtractor() {
+            return portExtractor;
+        }
+    }
+
+    private static final Module SMTP_PROTOCOL_MODULE = Modules.combine(
+            new ProtocolHandlerModule(),
+            new SMTPServerModule());
+
+    private final Host cassandraHost;
+    private final SmtpServerConnectedType smtpServerConnectedType;
+
+    private TemporaryFolder folder;
+    private GuiceJamesServer jamesServer;
+    private InMemoryDNSService inMemoryDNSService;
+    private ExternalSessionFactory sessionFactory;
+
+    public CassandraRabbitMQAwsS3SmtpTestRule(SmtpServerConnectedType smtpServerConnectedType, Host cassandraHost) {
+        this.smtpServerConnectedType = smtpServerConnectedType;
+        this.cassandraHost = cassandraHost;
+    }
+
+    @Override
+    public Statement apply(Statement base, Description description) {
+        return base;
+    }
+
+    @Override
+    public boolean addUser(String userAtDomain, String password) throws Exception {
+        Preconditions.checkArgument(userAtDomain.contains("@"), "The 'user' should contain the 'domain'");
+        Iterator<String> split = Splitter.on("@").split(userAtDomain).iterator();
+        split.next();
+        String domain = split.next();
```

Comment:
Manual parsing spotted!!!

Can we rely on `User` class as well?

And let's refactor that in other classes as well!